app/module/module.go

```diff
@@ -0,0 +1,349 @@
+			continue
+		}
+		ctx.Logger().Debug("running initialization for module", "module", moduleName)
+
+		moduleValUpdates := modules[moduleName].InitGenesis(ctx, cdc, genesisData[moduleName])
+
+		// use these validator updates if provided, the module manager assumes
+		// only one module will update the validator set
+		if len(moduleValUpdates) > 0 {
+			if len(validatorUpdates) > 0 {
+				panic("validator InitGenesis updates already set by a previous module")
+			}
+			validatorUpdates = moduleValUpdates
+		}
+	}
+
+	// a chain must initialize with a non-empty validator set
+	if len(validatorUpdates) == 0 {
+		panic(fmt.Sprintf("validator set is empty after InitGenesis, please ensure at least one validator is initialized with a delegation greater than or equal to the DefaultPowerReduction (%d)", sdk.DefaultPowerReduction))
+	}
+
+	return abci.ResponseInitChain{
+		Validators: validatorUpdates,
+	}
+}
+
+// ExportGenesis performs export genesis functionality for modules
+func (m *Manager) ExportGenesis(ctx sdk.Context, cdc codec.JSONCodec, version uint64) map[string]json.RawMessage {
+	genesisData := make(map[string]json.RawMessage)
+	modules := m.versionedModules[version]
+	for _, moduleName := range m.OrderExportGenesis {
+		genesisData[moduleName] = modules[moduleName].ExportGenesis(ctx, cdc)
+	}
+
+	return genesisData
+}
+
+// assertNoForgottenModules checks that we didn't forget any modules in the
+// SetOrder* functions.
+func (m *Manager) assertNoForgottenModules(setOrderFnName string, moduleNames []string) {
+	ms := make(map[string]bool)
+	for _, m := range moduleNames {
+		ms[m] = true
+	}
+	var missing []string
+	for _, m := range m.allModules {
+		if _, ok := ms[m.Name()]; !ok {
+			missing = append(missing, m.Name())
+		}
+	}
+	if len(missing) != 0 {
+		panic(fmt.Sprintf(
+			"%s: all modules must be defined when setting %s, missing: %v", setOrderFnName, setOrderFnName, missing))
+	}
+}
+
+// MigrationHandler is the migration function that each module registers.
+type MigrationHandler func(sdk.Context) error
+
+// VersionMap is a map of moduleName -> version
+type VersionMap map[string]uint64
+
+// RunMigrations performs in-place store migrations for all modules. This
+// function MUST be called when the state machine changes appVersion
+func (m Manager) RunMigrations(ctx sdk.Context, cfg sdkmodule.Configurator, fromVersion, toVersion uint64) error {
+	c, ok := cfg.(configurator)
+	if !ok {
+		return sdkerrors.ErrInvalidType.Wrapf("expected %T, got %T", configurator{}, cfg)
+	}
+	modules := m.OrderMigrations
+	if modules == nil {
+		modules = DefaultMigrationsOrder(m.ModuleNames(toVersion))
+	}
+	currentVersionModules, exists := m.versionedModules[fromVersion]
+	if !exists {
+		return sdkerrors.ErrInvalidVersion.Wrapf("version %d not supported", fromVersion)
+	}
+	nextVersionModules, exists := m.versionedModules[toVersion]
+	if !exists {
+		return sdkerrors.ErrInvalidVersion.Wrapf("version %d not supported", toVersion)
```

Comment:
[optional] to disambiguate from the error above

```suggestion
		return sdkerrors.ErrInvalidVersion.Wrapf("toVersion %d not supported", toVersion)
```